### PR TITLE
Changed Checks from TriviallyCopyable to TriviallyCopyConstructible 

### LINKED ISSUE #47355: range-loop-analysis checks for trivial copyability, rather than trivial copy-constructibility
|  |  |
| --- | --- |
| Bugzilla Link | [48011](https://llvm.org/bz48011) |
| Version | 11.0 |
| OS | Linux |
| CC | @dwblaikie,@devincoughlin,@randomnetcat,@pogo59,@zygoloid,@Cry-Thomas |

## Extended Description 
On the Cpplang slack, user Wolle posted a curious warning that came from range-loop-analysis.

Here's a minimal example: https://godbolt.org/z/1cxs43. Note that without the copy assignment operator overload, the warning is not present.

The range-based for loop performs a copy construction, so that is what should be checked for.

At https://clang.llvm.org/doxygen/SemaStmt_8cpp_source.html#l02824, the check is on whether the type is trivially copyable, not whether it is trivially copy constructible, which is a weaker requirement (but the relevant one in this case). I am not intimately familiar with clang sourcecode, but I believe this to be the cause of the warning.

### PR DESCRIPTION
**Pull request by 11happy**
**Overview:**
This pull request fixes #47355 where in the Clang compiler's range-loop-analysis incorrectly checks for trivial copyability instead of trivial copy constructibility, leading to erroneous warnings.

**Testing:**
- Tested the updated code.
- Verified that other functionalities remain unaffected.
Screenshot from 2024-01-01 19-50-20

**Dependencies:**
- No dependencies on other pull requests.

**References:**
- https://cplusplus.com/reference/type_traits/is_trivially_copy_constructible/
- https://en.cppreference.com/w/cpp/named_req/CopyConstructible
- https://cplusplus.com/reference/type_traits/is_trivially_copyable/

**CC:**
- @Endilll , @cor3ntin  , @AaronBallman

### CODE REVIEW

**Comment by 11happy** [comment]
Hello @cor3ntin In my previous pull request I mistakenly added add the commits so I closed that one and opened this new one.Sorry for any inconvenience .

**Comment by 11happy** [comment]
Hello @Endilll Can you please look into this and request for reviews and review this. I am sorry for my mistake.

File: clang/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -870,9 +870,9 @@ Bug Fixes to AST Handling
 - Fixed a bug where RecursiveASTVisitor fails to visit the
   initializer of a bitfield.
   `Issue 64916 <https://github.com/llvm/llvm-project/issues/64916>`_
-- Fixed a bug where Template Instantiation failed to handle Lambda Expressions
-  with certain types of Attributes.
-  (`#76521 <https://github.com/llvm/llvm-project/issues/76521>`_)
```
Comment (cor3ntin) [inline clang/docs/ReleaseNotes.rst:875]: Why did you remove that?

**Comment by cor3ntin** [comment]
Review previously done in https://github.com/llvm/llvm-project/pull/76680

File: clang/docs/ReleaseNotes.rst (hunk 1 above)
Comment (11happy) [reply clang/docs/ReleaseNotes.rst:875]: in my previous mistake I think it got removed unintentionally ,sorry from my end.

Comment (11happy) [reply clang/docs/ReleaseNotes.rst:875]: Should I add that to a new commit?

Comment (cor3ntin) [reply clang/docs/ReleaseNotes.rst]: Yes, please

Comment (11happy) [reply clang/docs/ReleaseNotes.rst]: corrected that.

**Review by cor3ntin — APPROVED** [review]
LGTM.
Congratulation on your first contribution!

**Comment by 11happy** [comment]
Thanks, I am really grateful for your guidance.

**Comment by 11happy** [comment]
Are there any more changes for this PR required from my end?
